ministryofjustice/laa-crime-means-assessment
# Lasb 759

## What

[LASB-759](https://dsdmoj.atlassian.net/browse/LASB-759)

Describe what you did and why.
Changes are as follow:
# Liquibase
I have added wiring and configured it to run run liquibase change sets on existing db. liquibase is disabled for testing. See:
- crime-means-assessment/src/main/resources/application.yaml -line 16 for liquibase changelog path
- crime-means-assessment/src/main/resources/db/changelog/db.changelog-master.yaml - for list of changes to perform to DB
- liquibase changes have the following format: **--changeset** which  indicates liquibase that the sql file is a liquibase changeset:
  - crime-means-assessment/src/main/resources/db/changelog/changeset/00-mla-create-schema.sql - schema creation.  **Note**: the name of the schema was chosen as a result of working on the MLA schema on the legacy DB. If a better name is suggested, I'm happy to change it 
  - crime-means-assessment/src/main/resources/db/changelog/changeset/01-defendant-assessment-table-create.sql - table creation
  - crime-means-assessment/src/main/resources/db/changelog/changeset/02-defendant-assessment-table-test-data-insert.sql - test data - this file is there only to insert test data
# RESTful API related functionality
## RestController 
It contains functionality for all CRUD operations. See:
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/controller/DefendantAssessmentController.java
## Service
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/service/DefendantAssessmentService.java
## Repository
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/repository/DefendantAssessmentRepository.java
## JPA Entity
 **Note**: the name of the schema was chosen as a result of working on the MLA schema on the legacy DB. If a better name is suggested, I'm happy to change it 
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/entity/DefendantAssessmentEntity.java
## Exception Handler functionality
###  @RestControllerAdvice
This class is used to handle all runtime exceptions encountered when calls are made to the RestController class. It returns the relevant HTTP status and it wraps the exception in a ErrorDTO object.
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/exceptions/DefentantAssessmentControllerExceptionHandler.java
#### Relevant exceptions thrown by the service that are processed by the DefentantAssessmentControllerExceptionHandler
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/exceptions/DefendantAssessmentIdPresentException.java
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/exceptions/DefendantAssessmentInvalidIdException.java
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/exceptions/DefendantAssessmentMissingException.java
- crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/defendant/exceptions/DefendantAssessmentNotFoundException.java

# Unit Testing
## Test Configuration
- I've modified some of the h2 DB related configuration values in order to successfully run unit tests.
  - see crime-means-assessment/src/test/resources/application.yaml 
- I've explicitly created a schema on h2 db because it is not doing it by default under the current configuration. I don't know how else to successfully build the DB.
  - see crime-means-assessment/src/test/resources/schema-h2.sql

I've created a helper class to have a few objects built and ready to help me run the tests. 
- see crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/data/builder/TestModelDataBuilder.java

## Controller testing
I've tried to cover every possible scenario I could think of when testing the Rest Controller, hence the long list of scenarios.
- see crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/defendant/controller/DefendantAssessmentControllerTest.java

## Service testing
- see crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/defendant/service/DefendantAssessmentServiceTest.java


## Repository Testing
Currently, it's a barebone repository testing. I just wanted to make sure the tables were being constructed and that data was being saved.
I am not sure it merits any more effort.
- see crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/defendant/repository/DefendantAssessmentRepositoryTest.java



## Checklist

Before you ask people to review this PR:

- [x] Tests should be passing: `./gradlew test`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase master`.
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against master and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.
